### PR TITLE
only score unscored moves if there are at least five scored ones

### PR DESCRIPTION
**Pull request by robertnurnberg**
Implements the fix discussed in https://github.com/vondele/cdbexplore/issues/50.

May need some testing, ideally in connection with the [extraStats branch](https://github.com/robertnurnberg/cdbexplore/tree/extraStats).

### CODE REVIEW

**Comment by robertnurnberg** [comment]
So this seems to work as expected. Depth 23 search on 1. g4 gives:
```
  queryall  :  2183437
  bf        :  1.89
  chessdbq  :  189201 (8.67% of queryall)
  enqueued  :  4123
  unscored  :  4 (0.10% of enqueued)
 EXTRA STATS: unscoredSiblings =  {5: 3, 9: 1}
  reprobed  :  2091 (1.11% of chessdbq)
 EXTRA STATS: reprobeGranted / reprobeAsked = 42 / 436 (0.10%)
```

BUT: We should wait for noob to come back to me on this discord [message](https://discord.com/channels/435943710472011776/1101022188313772083/1118582504140574841). If `queryall` stays as it is (and so does not provide 5 moves for positions beyond the old piece limit), then we have to do something on our side. Like automatically queuing nodes that come back with fewer than 5 moves. E.g. by doing a non-awaiting `queue` API call in [L306](https://github.com/vondele/cdbexplore/blob/8eb1453adf01db1e221b39b529bd42b6e0f05cec/cdbsearch.py#LL306C21-L306C21).

**Comment by robertnurnberg** [comment]
The situation is actually worse than I thought. Because for now we store incomplete move lists in the TT, meaning that they persist for longer than it takes cdb to complete them.

We could discuss a fix here or on discord. I guess one option is in `querall` to override `skipTT` if there are fewer than 5 scored moves, and in addition call `queue` when the API call `queryall` returns insufficient numbers of moves.

**Comment by robertnurnberg** [comment]
With the latest change I believe this is a possible (partial) solution.

- we only allow scoring unscored moves if we have a full set of scored moves from cdb
- for positions where cdb's move list is incomplete, we schedule `queue` in the background so that the list will hopefully be completed
- `querall` calls that lead to an incomplete move list being scored in our TT will be corrected automatically once `depth > depthForceQuery`

One issue that remains is the 3rd bullet point: at present positions with just a single scored move, say, need to wait until they are searched to depth 10 before our TT learns about the other four moves. I guess that is not really satisfactory. So in addition we may want to override `skipTT=False` to `True` in `querall` if move lists are incomplete? Then a possible downside would be repeated API calls in positions where cdb _never_ returns a full move list, something that noob hinted at on [discord](https://discord.com/channels/435943710472011776/1101022188313772083/1118768283403964416).

**Comment by robertnurnberg** [comment]
The latest commit should deal with bullet point 3.

**Comment by robertnurnberg** [comment]
So a quick summary what this PR fixes.

In current master/main we score far too many `unscored` moves. The main cause of this seems to be that cdb's `queryall` API can (unexpectedly) return incomplete move lists, i.e. a list of scored moves that is shorter than 5 even though the position has more legal moves. This is evidenced by the example given in https://github.com/vondele/cdbexplore/issues/50#issuecomment-1591448821, where `EXTRA STATS: unscoredSiblings =  {1: 378, 2: 6, 5: 18, 6: 2, 8: 2, 10: 2}` means that of the 408 `unscored` moves, 378 were from situations where `queryall` returned only a single scored move.

This has quite dramatic and unforeseen consequences. For starters, we save these too short move lists into our local TT, where they will remain untouched until `depth > depthForceQuery` leads to a re-query of the position. Now the first `depth=2` search of a position with a 1-move list will inevitably lead to the scoring of an `unscored` move, since `move_depth` will return `min(0, depth - 2)`. Depending how bad the randomly picked move is, the same could happen again on later searches, with a 2-move list and at higher depth. (In the above example this happened 6 times overall.) But even if no more `unscored` moves are scored in this position, until `depth > depthForceQuery` the move list will be stuck with basically just one good move. This is arguably a bug in the current implementation.

This PR fixes the described shortcomings in the following way:

- We only allow scoring `unscored` moves if we already have a complete move list from cdb.
- For nodes where we still do not have a complete move list from cdb, we schedule a `queue` API in the background, since according to noob `queryall` alone is not enough to guarantee complete move lists for positions beyond the old piece count limit.
- In addition, for such nodes, we also populate `skipTT_db_moves` by calling our `queryall` method with `skipTT=True`. In this way, we immediately benefit from longer move lists becoming available on cdb, either due to our earlier queries, or due to our launched background `queue` calls.

A worry may be that we completely forbid to score `unscored` moves when the cdb move list is not complete. But here I would simply trust cdb. In fact, noob has confirmed on [discord](https://discord.com/channels/435943710472011776/1101022188313772083/1118900671232614531) that in general there should always be at least 5 moves in the scored move list, once `queue` has done its work.

I think this is ready to be merged.

File: cdbsearch.py [hunk 1]
```diff
@@ -373,17 +376,22 @@ async def search(self, board, depth):
...
 
-        scoreCount = len(scored_db_moves) - 1  # number of scored moves for board
+        scoredCount = len(scored_db_moves) - 1  # number of scored moves for board
+        movesCount = len(list(board.legal_moves))  # numer of legal moves
+        missingCDBmoves = scoredCount < min(CDB_SIEVED, movesCount)
+
+        # for positions with an incomplete move list, we schedule a queue API call, since querall for positions beyond cdb's old piece count limit may not be enough to add new moves
+        if missingCDBmoves:
+            asyncio.ensure_future(
+                self.__cdbapicall(f"?action=queue&board={epd}&json=1", timeout=60)
```
Comment (vondele) [inline cdbsearch.py:386]: there is no counting of these queueing calls.

**Comment by vondele** [comment]
once that review comment is addressed this can be merged IMO.

**Comment by robertnurnberg** [comment]
Thanks for the review. I will merge this now, as a "bug"-fix, and then we can deal with the counting in a follow-up PR.

I sort of left the counting out on purpose, because (a) these calls do not fit into any of the existing categories, (b) they are counted for the inflightR average, and (c) they are not guaranteed to lead to some action by cdb.

**Pull request merged**